### DISCUSSION TITLE
Currently not moving to openpdf 2.X as it requires java 17

### DISCUSSION THREAD

**Discussion by fugerit79** [fugerit-org/fj-doc#128] [General] [answered]
As stated on release : 
https://github.com/LibrePDF/OpenPDF/releases/tag/2.0.0

Currently we are stuck on openpdf 1.x for backward compatibility.

**Comment by andreasrosdalw** [comment]
# Advantages of Moving from Java 8 to Java 17+

Moving from Java 8 to Java 17+ offers several advantages:

1. **Performance Improvements**: Each new version of Java typically comes with performance enhancements, including optimizations in the runtime environment and improvements in the efficiency of the Java Virtual Machine (JVM). This can lead to better overall application performance and responsiveness.

2. **Language Enhancements**: Java has evolved significantly since Java 8, with new language features introduced in subsequent versions. For example, Java 9 introduced modularity with the Java Platform Module System (JPMS), Java 10 introduced local-variable type inference (var keyword), Java 11 introduced the HTTP client API, Java 12 introduced switch expressions, Java 14 introduced Records, and Java 16 introduced sealed classes. These language enhancements can lead to more concise, readable, and maintainable code.

3. **Security Improvements**: Newer versions of Java often come with security enhancements and bug fixes, addressing vulnerabilities found in earlier versions. Upgrading to the latest version helps ensure that your applications are more secure and less susceptible to attacks.

4. **New APIs and Libraries**: Each Java release typically introduces new APIs and libraries, offering developers access to improved functionality and tools for building applications. For example, Java 9 introduced the Flow API for reactive programming, Java 11 introduced the HttpClient API for making HTTP requests, and Java 14 introduced the Pattern Matching API.

5. **Better Tooling Support**: IDEs and other development tools tend to offer better support for newer versions of Java. Upgrading to a newer version can provide access to improved tooling features, making development and debugging easier and more efficient.

6. **Long-term Support**: Starting from Java 11, Oracle introduced a new release cadence with long-term support (LTS) releases every three years. This means that versions like Java 11, Java 17, Java 23, etc., receive long-term support, including security updates and bug fixes, for a longer duration, making them more suitable for enterprise use.

7. **Compatibility with Modern Libraries and Frameworks**: Many modern libraries and frameworks require a minimum Java version beyond Java 8. Upgrading to a newer version of Java enables you to leverage the latest features and improvements in these libraries and frameworks, ensuring compatibility with the broader ecosystem of Java tools and technologies.

8. **Community Support**: As newer versions of Java gain traction, community support for older versions like Java 8 may decline over time. Upgrading to a newer version ensures that you can benefit from ongoing community support, including access to tutorials, forums, and open-source contributions.

Overall, upgrading from Java 8 to Java 17+ can result in improved performance, enhanced security, access to new language features and APIs, better tooling support, and compatibility with modern libraries and frameworks, ultimately leading to better maintainability and scalability of your Java applications.

**Reply by mttfranci** [reply]
Yes of course, there are all very good reasons to move to java 17+. 
In most project where this framework is included, java 17 or 21 is used.
But are there is still some project using Java 8, we are sticking to it for a while for source compatibility.
To use java 21 in runtime project still get you the change of getting most advantages and at the same time have good backward compatibility.

Maybe we can add a new module based on openpdf 2.0.2 and keeping both in the project?

**Comment by fugerit79** [comment, marked as answer]
Now OpenPDF version is loaded in dynamic way : 

<https://github.com/fugerit-org/fj-doc/blob/main/CHANGELOG.md#8160---2025-08-22>

If using java 21+ latest opendf version is selected
